src/forest.rs

```diff
@@ -243,10 +241,324 @@ impl Forest {
+        assert_eq!(id, 1);
+        assert_eq!(node.style, s2);
+        assert!(node_measure_eq(&node, measure_fn2));
+        assert_forest_size(&forest, 2);
+    }
+
+    #[test]
+    fn new_with_children_single() {
+        let mut forest = Forest::with_capacity(2);
+        let style = FlexboxLayout { flex_grow: 1.0, ..Default::default() };
+        let c1_id = add_default_leaf(&mut forest);
+        let mut children = new_vec_with_capacity(1);
+        children.push(c1_id);
+
+        let id = forest.new_with_children(style, children);
+        let new_node = &forest.nodes[id];
+
+        assert_eq!(id, 1);
+        assert_eq!(new_node.style, style);
+        assert_eq!(forest.parents[c1_id][0], id);
+        assert_eq!(forest.children[id][0], c1_id);
+        assert_forest_size(&forest, 2);
+    }
+
+    #[test]
+    fn new_with_children_multiple() {
+        let mut forest = Forest::with_capacity(3);
+        let style = FlexboxLayout { flex_grow: 1.0, ..Default::default() };
+        let c1_id = add_default_leaf(&mut forest);
+        let c2_id = add_default_leaf(&mut forest);
+        let mut children = new_vec_with_capacity(2);
+        children.push(c1_id);
+        children.push(c2_id);
+
+        let id = forest.new_with_children(style, children);
+        let new_node = &forest.nodes[id];
+
+        assert_eq!(id, 2);
+        assert_eq!(new_node.style, style);
+        assert_eq!(forest.parents[c1_id][0], id);
+        assert_eq!(forest.parents[c2_id][0], id);
+        assert_eq!(forest.children[id][0], c1_id);
+        assert_eq!(forest.children[id][1], c2_id);
+        assert_forest_size(&forest, 3);
+    }
+
+    #[test]
+    fn add_child() {
+        let mut forest = Forest::with_capacity(2);
+        let parent_id = add_default_leaf(&mut forest);
+        let child_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, child_id);
+
+        let parent = &forest.nodes[parent_id];
+
+        assert_eq!(forest.parents[child_id][0], parent_id);
+        assert_eq!(forest.children[parent_id][0], child_id);
+        assert!(parent.is_dirty);
+    }
+
+    #[test]
+    fn add_child_multiple() {
+        let mut forest = Forest::with_capacity(3);
+        let parent_id = add_default_leaf(&mut forest);
+        let c1_id = add_default_leaf(&mut forest);
+        let c2_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, c1_id);
+        forest.add_child(parent_id, c2_id);
+
+        let parent = &forest.nodes[parent_id];
+
+        assert_eq!(forest.parents[c1_id][0], parent_id);
+        assert_eq!(forest.parents[c2_id][0], parent_id);
+        assert_eq!(forest.children[parent_id][0], c1_id);
+        assert_eq!(forest.children[parent_id][1], c2_id);
+        assert!(parent.is_dirty);
+    }
+
+    #[test]
+    fn clear() {
+        let mut forest = Forest::with_capacity(1);
+        add_default_leaf(&mut forest);
+        forest.clear();
+        assert_forest_size(&forest, 0);
+    }
```

Comment:
Does adding an assert after the setup and before calling the methods to test help convey more precisely what is being tested?

```suggestion
    fn clear() {
        let mut forest = Forest::with_capacity(1);
        add_default_leaf(&mut forest);
        assert_forest_size(&forest, 1); // <- Before something changes
        forest.clear();
        assert_forest_size(&forest, 0); // <- This should be the result of the change
    }
```

I personally like doing this and I find it help with conveying _intent_, but it might work best on very narrow tests like this and be more confusing on tests with many asserts to capture the full picture

Reply:
While I do think that this make the intent clearer in this case, I'm a bit against it. I see some problems by doing this:
- `new_leaf` is already tested and will fail if it doesn't do what it's supposed to do.
- If this test fails because `new_leaf` behaved badly we'll have two broken tests, instead of just one which would be telling us exactly what went wrong.
- And personally it bothers me a little that it breaks the Arrange, Act, Assert pattern.
All of these points also apply to other test cases, not only this `clear` one.

But anyway, if you think this would make our test suite more standardized, let me know and I'll change it :)

Reply:
I agree with your stance here @mcrvaz. Being able to isolate exactly what failed from unit tests is incredibly valuable.